src/app/util/ember-compatibility-functions.cpp

```diff
@@ -250,6 +250,84 @@ CHIP_ERROR SendFailureStatus(const ConcreteAttributePath & aPath, AttributeRepor
     return aAttributeReport.EndOfAttributeReportIB().GetError();
 }
 
+// This reader should never actually be registered; we do manual dispatch to it
+// for the one attribute it handles.
+class MandatoryGlobalAttributeReader : public AttributeAccessInterface
+{
+public:
+    MandatoryGlobalAttributeReader(const EmberAfCluster * aCluster) :
+        AttributeAccessInterface(MakeOptional(kInvalidEndpointId), kInvalidClusterId), mCluster(aCluster)
+    {}
+
+protected:
+    const EmberAfCluster * mCluster;
+};
+
+class AttributeListReader : public MandatoryGlobalAttributeReader
+{
+public:
+    AttributeListReader(const EmberAfCluster * aCluster) : MandatoryGlobalAttributeReader(aCluster) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+};
+
+CHIP_ERROR AttributeListReader::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    // The id of AttributeList is not in the attribute metadata.
+    // TODO: This does not play nicely with wildcard reads.  Need to fix ZAP to
+    // put it in the metadata, or fix wildcard expansion to add it.
+    return aEncoder.EncodeList([this](const auto & encoder) {
+        constexpr AttributeId ourId = Clusters::Globals::Attributes::AttributeList::Id;
+        const size_t count          = mCluster->attributeCount;
+        bool addedOurId             = false;
+        for (size_t i = 0; i < count; ++i)
+        {
+            AttributeId id = mCluster->attributes[i].attributeId;
+            if (!addedOurId && id > ourId)
```

Comment:
Is this essentially future-proofing for when the metadata does get added `AttributeList`?

Reply:
Why `>`? Why not `==`?

Reply:
What this code is doing is inserting `ourId` in the (sorted) list in the right spot.  So we want to insert it before the first value that is larger than `ourId`.

`==` would never test true, since the whole point is that `ourId` is not in the list.

The `addedOurId` check is because there are multiple ids larger than ours, and we don't want to double-add.

Reply:
Gotcha. I had mis-read the intention, but given that we both kinda agree about not adding this to the metadata, we don't need to future-proof for when the metadata does show-up (since it likely never will).

Is the assumption that `attributes` is sorted well defined? I guess if it isn't, this will still only insert this global attribute once, and in a pseudo-reasonable location.

Reply:
For the generated metadata, the sorting assumption is true.

We _could_ just append this attribute at the end, but the intent was to make it look "normal" and that is sorted.